pulp_smash/cli.py

```diff
@@ -184,6 +190,97 @@ def run(self, args, **kwargs):
         completed_process = CompletedProcess(args, code, stdout, stderr)
         return self.response_handler(completed_process)
+
+
+class Service(object):
+    """A service on a system.
+
+    Each instance of this class represents a single service running on a single
+    system. An example may help to clarify this idea:
+
+    >>> from pulp_smash import cli, config
+    >>> service = cli.Service(config.get_config(), 'httpd')
+    >>> completed_process = service.stop()
+    >>> completed_process = service.start()
+
+    In the example above, the ``service`` object represents the "httpd" service
+    on the host referenced by :func:`pulp_smash.config.get_config`.
+
+    Upon instantiation, a :class:`Service` object talks to its target
+    system and uses simple heuristics to determine which type of service
+    manager is used (SysV or systemd). As a result, it's possible to manage
+    services on significantly different systems with exactly the same commands:
+
+    >>> from pulp_smash import cli
+    >>> from pulp_smash.config import ServerConfig
+    >>> configs = ServerConfig().read('old'), ServerConfig().read('new')
+    >>> services = (cli.Service(config, 'httpd') for config in configs)
+    >>> [service.start().args for service in services]
+    'service httpd start'
+    'systemctl start httpd'
+    """
+
+    def __init__(self, server_config, service):
+        """Initialize a new object."""
+        self._client = Client(server_config)
+        self._command_builder = None
+
+        # Set `self._command_builder`.
+        # FIXME: Conditionally use sudo. Explain when sudo is used.
+        service_manager = self._get_service_manager(server_config)
+        if service_manager == 'systemd':
+            self._command_builder = lambda verb: (
+                'sudo', 'systemctl', verb, service
+            )
+        elif service_manager == 'sysv':
+            self._command_builder = lambda verb: (
+                'sudo', 'service', service, verb
+            )
+        assert self._command_builder is not None
+
+    @staticmethod
+    def _get_service_manager(server_config):
+        """Talk to the target system and determine the type of service manager.
+
+        Return "systemd" or "sysv" if the service manager appears to be one of
+        those. Raise an exception otherwise.
+        """
+        hostname = _get_hostname(server_config.base_url)
+        try:
+            return _SERVICE_MANAGERS[hostname]
+        except KeyError:
+            pass
+
+        client = Client(server_config)
+        for command, svc_manager in (
+                (('which', 'systemctl'), 'systemd'),
+                (('which', 'service'), 'sysv'),
+        ):
+            try:
+                client.run(command)
+                _SERVICE_MANAGERS[hostname] = svc_manager
+                return svc_manager
+            except subprocess.CalledProcessError:
+                pass
+        raise Exception  # FIXME: Raise a custom exception.
```

Comment:
There is no single `pulp_smash.exceptions` module or anything like it. They're sprinkled throughout the codebase. :/